java/src/main/java/com/saucelabs/saucebindings/SauceSession.java

```diff
@@ -34,54 +38,87 @@ public SauceSession(BaseConfigurations configs) {
+            setVisual();
+        }
+
+        if (isVisualSession && sauceOptions.sauce().getName() == null) {
+            String msg = "Visual Tests Require setting a name in options: SauceOptions#setName(Name)";
+            throw new InvalidSauceOptionsArgumentException(msg);
+        }
         driver = createRemoteWebDriver(getSauceUrl(), sauceOptions.toCapabilities());
+
+        if (isVisualSession) {
+            visualSession.start(driver, sauceOptions.sauce().getName());
+        }
+
         return driver;
 	}
 
     public URL getSauceUrl() {
-        if (sauceUrl != null) {
-            return sauceUrl;
-        } else {
-            try {
+        try {
+            if (sauceUrl != null) {
+                return sauceUrl;
+            } else {
                 return new URL(dataCenter.getValue());
-            } catch (MalformedURLException e) {
-                throw new InvalidArgumentException("Invalid URL");
             }
+        } catch (MalformedURLException e) {
+            throw new InvalidArgumentException("Invalid URL");
         }
     }
 
     protected RemoteWebDriver createRemoteWebDriver(URL url, MutableCapabilities capabilities) {
         return new RemoteWebDriver(url, capabilities);
     }
 
+    public VisualSession visual() {
+        return visualSession;
+    }
+
     public void stop(Boolean passed) {
         String result = passed ? "passed" : "failed";
         stop(result);
     }
 
     public void stop(String result) {
-        updateResult(result);
-        stop();
+        if (this.driver != null) {
+            updateResult(result);
+            stop();
+        }
     }
 
     private void updateResult(String result) {
         getDriver().executeScript("sauce:job-result=" + result);
+
         // Add output for the Sauce OnDemand Jenkins plugin
         // The first print statement will automatically populate links on Jenkins to Sauce
         // The second print statement will output the job link to logging/console
-        if (this.driver != null) {
-            String sauceReporter = String.format("SauceOnDemandSessionID=%s job-name=%s", this.driver.getSessionId(), this.sauceOptions.sauce().getName());
-            String sauceTestLink = String.format("Test Job Link: https://app.saucelabs.com/tests/%s", this.driver.getSessionId());
-            System.out.print(sauceReporter + "\n" + sauceTestLink + "\n");
-        }
+        String sauceReporter = String.format("SauceOnDemandSessionID=%s job-name=%s",
+                this.driver.getSessionId(),
+                this.sauceOptions.sauce().getName());
+        String sauceTestLink = String.format("Test Job Link: https://app.saucelabs.com/tests/%s",
+                this.driver.getSessionId());
+        System.out.print(sauceReporter + "\n" + sauceTestLink + "\n");
     }
 
     private void stop() {
-        if(driver !=null) {
-            driver.quit();
+        if (isVisualSession) {
```

Comment:
Shouldn't `stop()` also call `executeScript("/*@visual.end*/")`?

Reply:
That's what the following line of code does when it prints (`visual().getResults()`)